File: frontend/src/lib/api/schemas.ts
import { z } from 'zod';

const hex32 = z.string().regex(/^0x[a-fA-F0-9]{64}$/, 'expected 0x + 64 hex');
const address = z.string().regex(/^0x[a-fA-F0-9]{40}$/, 'expected 0x + 40 hex');

export const Confidence = z.enum(['High', 'Medium', 'Low', 'Unverified']);
export type Confidence = z.infer<typeof Confidence>;

export const Evidence = z.object({
  title: z.string(),
  source: z.string(),
  url: z.string().optional().default(''),
  excerpt: z.string().optional().default(''),
  claim: z.string().optional().default(''),
  confidence: Confidence,
  hash: hex32,
});
export type Evidence = z.infer<typeof Evidence>;

export const AnalyzeRequest = z.object({
  productName: z.string().min(1),
  productUrl: z.string().optional(),
  brand: z.string().optional(),
});
export type AnalyzeRequest = z.infer<typeof AnalyzeRequest>;

export const AnalyzeResponse = z.object({
  productName: z.string(),
  brand: z.string().optional().default(''),
  score: z.number().int().min(0).max(100),
  grade: z.string(),
  summary: z.string().optional().default(''),
  findings: z.array(z.string()).default([]),
  evidences: z.array(Evidence).default([]),
  reportHash: hex32,
  evidenceMerkleRoot: hex32,
  reportURI: z.string(),
  metadataURI: z.string(),
});
export type AnalyzeResponse = z.infer<typeof AnalyzeResponse>;

export const MintRequest = z.object({
  to: address,
  productName: z.string().min(1),
  brand: z.string(),
  score: z.number().int().min(0).max(100),
  grade: z.string(),
  reportHash: hex32,
  evidenceMerkleRoot: hex32,
  metadataURI: z.string(),
});
export type MintRequest = z.infer<typeof MintRequest>;

export const MintResponse = z.object({
  tokenId: z.union([z.number(), z.string()]).transform((v) => BigInt(v)),
  transactionHash: hex32,
  contractAddress: address,
});
export type MintResponse = z.infer<typeof MintResponse>;

export const AnalyzeAndMintRequest = AnalyzeRequest.extend({ to: address });
export type AnalyzeAndMintRequest = z.infer<typeof AnalyzeAndMintRequest>;

export const AnalyzeAndMintResponse = AnalyzeResponse.merge(
  MintResponse.pick({ tokenId: true, transactionHash: true, contractAddress: true }),
);
export type AnalyzeAndMintResponse = z.infer<typeof AnalyzeAndMintResponse>;

export const FullReport = z.object({
  productName: z.string(),
  brand: z.string().optional().default(''),
  score: z.number().int().min(0).max(100),
  grade: z.string(),
  summary: z.string().optional().default(''),
  positiveSignals: z.array(z.string()).optional().default([]),
  riskSignals: z.array(z.string()).optional().default([]),
  greenwashingRisk: z.string().optional().default(''),
  findings: z.array(z.string()).default([]),
  evidences: z.array(Evidence).default([]),
  alternatives: z.array(z.string()).optional().default([]),
  createdAt: z.string().optional(),
});
export type FullReport = z.infer<typeof FullReport>;

export const GetReportResponse = z.object({
  reportHash: hex32,
  reportURI: z.string(),
  report: FullReport,
});
export type GetReportResponse = z.infer<typeof GetReportResponse>;

File: backend/app/models/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


# ── Evidence ──────────────────────────────────────────────────────────────────

class Evidence(BaseModel):
    title: str
    source: str
    url: str = ""
    excerpt: str
    claim: str
    confidence: Literal["High", "Medium", "Low", "Unverified"] = "Medium"
    hash: str = ""  # populated after hashing


# ── Report ────────────────────────────────────────────────────────────────────

class GreenReport(BaseModel):
    productName: str
    brand: str
    score: int = Field(ge=0, le=100)
    grade: str
    summary: str
    positiveSignals: list[str] = []
    riskSignals: list[str] = []
    greenwashingRisk: Literal["Low", "Medium", "High", "Critical"] = "Medium"
    findings: list[str] = []
    evidences: list[Evidence] = []
    alternatives: list[str] = []
    createdAt: str = Field(default_factory=lambda: datetime.utcnow().isoformat())


# ── Request bodies ────────────────────────────────────────────────────────────

class AnalyzeRequest(BaseModel):
    productName: str = Field(..., min_length=1, examples=["Nike Pegasus Trail 5 DV3865-602"])
    productUrl: str = ""
    brand: str = ""


class MintRequest(BaseModel):
    to: str = Field(..., pattern=r"^0x[0-9a-fA-F]{40}$")
    productName: str
    brand: str
    score: int = Field(ge=0, le=100)
    grade: str
    reportHash: str = Field(..., pattern=r"^0x[0-9a-fA-F]{64}$")
    evidenceMerkleRoot: str = Field(..., pattern=r"^0x[0-9a-fA-F]{64}$")
    metadataURI: str


class AnalyzeAndMintRequest(BaseModel):
    productName: str = Field(..., min_length=1, examples=["Nike Pegasus Trail 5 DV3865-602"])
    productUrl: str = ""
    brand: str = ""
    to: str = Field(..., pattern=r"^0x[0-9a-fA-F]{40}$")


# ── Response bodies ───────────────────────────────────────────────────────────

class AnalyzeResponse(BaseModel):
    productName: str
    brand: str
    score: int
    grade: str
    summary: str
    findings: list[str]
    evidences: list[Evidence]
    reportHash: str
    evidenceMerkleRoot: str
    reportURI: str
    metadataURI: str


class MintResponse(BaseModel):
    tokenId: int
    transactionHash: str
    contractAddress: str


class AnalyzeAndMintResponse(BaseModel):
    # report fields
    productName: str
    brand: str
    score: int
    grade: str
    summary: str
    findings: list[str]
    evidences: list[Evidence]
    reportHash: str
    evidenceMerkleRoot: str
    reportURI: str
    metadataURI: str
    # mint fields
    tokenId: int
    transactionHash: str
    contractAddress: str


class ReportLookupResponse(BaseModel):
    reportHash: str
    reportURI: str
    report: GreenReport


class ErrorResponse(BaseModel):
    detail: str
    code: str = "INTERNAL_ERROR"
